File: desktop/tests/Qlipq.Ffmpeg.Tests/ProbeTests.cs
using Qlipq.Core;
using Qlipq.Ffmpeg;
using Xunit;

namespace Qlipq.Ffmpeg.Tests;

public class ProbeTests
{
    [Fact]
    public void BuildProbeArgsRequestsJsonFormatAndStreams()
    {
        Assert.Equal(
            ["-v", "error", "-print_format", "json", "-show_format", "-show_streams", "clip.mkv"],
            Probe.BuildProbeArgs("clip.mkv"));
    }

    [Fact]
    public void ParseFrameRateHandlesRationalsAndIntegers()
    {
        Assert.Equal(29.97, Probe.ParseFrameRate("30000/1001"));
        Assert.Equal(60, Probe.ParseFrameRate("60/1"));
        Assert.Equal(0, Probe.ParseFrameRate("0/0"));
        Assert.Equal(0, Probe.ParseFrameRate(null));
    }

    [Fact]
    public void ParseFfprobeExtractsVideoAndAudioRelativeIndices()
    {
        var probe = new FfprobeOutput
        {
            Streams =
            [
                new FfprobeStream { Index = 0, CodecType = "video", CodecName = "h264", Width = 2560, Height = 1440, RFrameRate = "60/1" },
                new FfprobeStream { Index = 1, CodecType = "audio", CodecName = "aac", Channels = 2, Tags = new() { ["title"] = "Desktop" } },
                new FfprobeStream { Index = 2, CodecType = "audio", CodecName = "aac", Channels = 1, Tags = new() { ["language"] = "eng", ["title"] = "Mic" } },
            ],
            Format = new FfprobeFormat { Duration = "63.500000", Size = "104857600" },
        };
        var info = Probe.ParseFfprobe(probe);
        Assert.Equal(63.5, info.DurationSec);
        Assert.Equal(2560, info.Width);
        Assert.Equal(1440, info.Height);
        Assert.Equal("h264", info.VideoCodec);
        Assert.Equal(60, info.Fps);
        Assert.Equal(104857600, info.SizeBytes);
        Assert.Equal(
            [
                new AudioStreamInfo { StreamIndex = 1, Index = 0, Codec = "aac", Channels = 2, Language = null, Title = "Desktop" },
                new AudioStreamInfo { StreamIndex = 2, Index = 1, Codec = "aac", Channels = 1, Language = "eng", Title = "Mic" },
            ],
            info.AudioStreams);
    }

    [Fact]
    public void ParseFfprobeAcceptsAJsonString()
    {
        var info = Probe.ParseFfprobe("{\"streams\":[],\"format\":{\"duration\":\"1.0\"}}");
        Assert.Equal(1, info.DurationSec);
        Assert.Empty(info.AudioStreams);
    }
}

File: iced/crates/qlipq-ffmpeg/tests/probe_tests.rs
use std::collections::HashMap;

use qlipq_core::media::AudioStreamInfo;
use qlipq_ffmpeg::probe::*;

fn close(a: f64, b: f64) -> bool {
    (a - b).abs() < 1e-9
}

#[test]
fn build_probe_args_requests_json() {
    assert_eq!(
        build_probe_args("clip.mkv"),
        vec!["-v", "error", "-print_format", "json", "-show_format", "-show_streams", "clip.mkv"]
    );
}

#[test]
fn parse_frame_rate_handles_rationals_and_integers() {
    assert!(close(parse_frame_rate(Some("30000/1001")), 29.97));
    assert!(close(parse_frame_rate(Some("60/1")), 60.0));
    assert!(close(parse_frame_rate(Some("0/0")), 0.0));
    assert!(close(parse_frame_rate(None), 0.0));
}

#[test]
fn parse_ffprobe_extracts_video_and_audio_indices() {
    let mut t1 = HashMap::new();
    t1.insert("title".to_string(), "Desktop".to_string());
    let mut t2 = HashMap::new();
    t2.insert("language".to_string(), "eng".to_string());
    t2.insert("title".to_string(), "Mic".to_string());

    let data = FfprobeOutput {
        streams: Some(vec![
            FfprobeStream { index: 0, codec_type: Some("video".into()), codec_name: Some("h264".into()), width: Some(2560), height: Some(1440), r_frame_rate: Some("60/1".into()), ..Default::default() },
            FfprobeStream { index: 1, codec_type: Some("audio".into()), codec_name: Some("aac".into()), channels: Some(2), tags: Some(t1), ..Default::default() },
            FfprobeStream { index: 2, codec_type: Some("audio".into()), codec_name: Some("aac".into()), channels: Some(1), tags: Some(t2), ..Default::default() },
        ]),
        format: Some(FfprobeFormat { duration: Some("63.500000".into()), size: Some("104857600".into()) }),
    };
    let info = parse_ffprobe_output(&data);
    assert!(close(info.duration_sec, 63.5));
    assert_eq!(info.width, 2560);
    assert_eq!(info.height, 1440);
    assert_eq!(info.video_codec, "h264");
    assert!(close(info.fps, 60.0));
    assert_eq!(info.size_bytes, Some(104857600));
    assert_eq!(
        info.audio_streams,
        vec![
            AudioStreamInfo { stream_index: 1, index: 0, codec: "aac".into(), channels: 2, language: None, title: Some("Desktop".into()) },
            AudioStreamInfo { stream_index: 2, index: 1, codec: "aac".into(), channels: 1, language: Some("eng".into()), title: Some("Mic".into()) },
        ]
    );
}

#[test]
fn parse_ffprobe_accepts_json_string() {
    let info = parse_ffprobe(r#"{"streams":[],"format":{"duration":"1.0"}}"#);
    assert!(close(info.duration_sec, 1.0));
    assert!(info.audio_streams.is_empty());
}
